dll/win32/shell32/wine/shellpath.c

```diff
@@ -474,42 +595,118 @@ int WINAPI PathCleanupSpec( LPCWSTR lpszPathW, LPWSTR lpszFileW )
 }
 
 /*************************************************************************
- * PathQualifyA		[SHELL32]
+ * PathQualifyW		[SHELL32]
  */
-static BOOL PathQualifyA(LPCSTR pszPath)
+static VOID PathQualifyW(LPWSTR pszPath)
 {
-	FIXME("%s\n",pszPath);
-	return FALSE;
+    TRACE("%s\n",debugstr_w(pszPath));
+    PathQualifyExW(pszPath, NULL, 0);
 }
 
 /*************************************************************************
- * PathQualifyW		[SHELL32]
+ * PathQualifyA		[SHELL32]
  */
-static BOOL PathQualifyW(LPCWSTR pszPath)
+static VOID PathQualifyA(LPSTR pszPath)
 {
-	FIXME("%s\n",debugstr_w(pszPath));
-	return FALSE;
+    WCHAR szPath[MAX_PATH];
+    TRACE("%s\n",pszPath);
+    SHAnsiToUnicode(pszPath, szPath, _countof(szPath));
+    PathQualifyW(szPath);
+    SHUnicodeToAnsi(szPath, pszPath, MAX_PATH);
 }
 
 /*************************************************************************
  * PathQualify	[SHELL32.49]
  */
-BOOL WINAPI PathQualifyAW(LPCVOID pszPath)
+VOID WINAPI PathQualifyAW(LPVOID pszPath)
 {
-	if (SHELL_OsIsUnicode())
-	  return PathQualifyW(pszPath);
-	return PathQualifyA(pszPath);
+    if (SHELL_OsIsUnicode())
+        PathQualifyW(pszPath);
+    else
+        PathQualifyA(pszPath);
 }
 
-static BOOL PathResolveA(LPSTR path, LPCSTR *paths, DWORD flags)
+BOOL WINAPI PathResolveA(LPSTR path, LPCSTR *dirs, DWORD flags)
 {
-    FIXME("(%s,%p,0x%08x),stub!\n", debugstr_a(path), paths, flags);
+    FIXME("(%s,%p,0x%08x),stub!\n", debugstr_a(path), dirs, flags);
     return FALSE;
 }
 
-static BOOL PathResolveW(LPWSTR path, LPCWSTR *paths, DWORD flags)
+BOOL WINAPI PathResolveW(LPWSTR path, LPCWSTR *dirs, DWORD flags)
 {
-    FIXME("(%s,%p,0x%08x),stub!\n", debugstr_w(path), paths, flags);
+    DWORD dwWhich;
+    TRACE("PathResolveW(%s,%p,0x%08x)\n", debugstr_w(path), dirs, flags);
+
+    dwWhich = ((flags & PRF_DONTFINDLNK) ? 0xF : 0x3F);
```

Comment:
What are these hardcoded values?

Reply:
This is the `dwWhich` parameter of both `shlwapi!PathFindOnPathExW` and `shlwapi!PathFileExistsDefExtW`. That represents a combination of the filename extension types.

Reply:
I see... Do you know the human-readable names of these values (and then use OR-ed flags instead)? (If this makes sense)

Reply:
I will do.

Reply:
OK.